[SPARK-28573][SQL] Convert InsertIntoTable(HiveTableRelation) to DataSource inserting for partitioned table

advancedxy opened the pull request:
## What changes were proposed in this pull request?
Datasource table now supports partition tables long ago. This commit adds the ability to translate
the InsertIntoTable(HiveTableRelation) to datasource table insertion.

## How was this patch tested?
Existing tests with some modification

advancedxy committed:
[SPARK-28573][SQL] Convert InsertIntoTable(HiveTableRelation) to Datasource inserting for partitioned table

AmplabJenkins commented:
Can one of the admins verify this patch?

advancedxy commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -197,7 +197,9 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
 
         val logicalRelation = cached.getOrElse {
-          val sizeInBytes = relation.stats.sizeInBytes.toLong
+          val defaultSizeInBytes = sparkSession.sessionState.conf.defaultSizeInBytes
```
For write path, the stats for HiveTable is not calculated as InsertIntoTable's table filed is not added in the children.

So `org.apache.spark.sql.hive.DetermineTableStats` skip this table

cloud-fan replied:
cc @wangyum do you think this is fixable?

wangyum replied:
Maybe we can be recalculate stats by:
https://github.com/apache/spark/blob/485ae6d1818e8756a86da38d6aefc8f1dbde49c2/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/HadoopFsRelation.scala#L74-L77

This is test case:
https://github.com/apache/spark/blob/485ae6d1818e8756a86da38d6aefc8f1dbde49c2/sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala#L1488-L1500

advancedxy replied:
Fixed by DetermineTableStats handling InsertIntoTable(HiveTableRelation, ...)

advancedxy commented:
cc @cloud-fan

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #108463 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108463/testReport)** for PR 25306 at commit [`220a9fb`](https://github.com/apache/spark/commit/220a9fb0cc854bf489e820e727af1b0de82958b8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13558/
Test PASSed.

cloud-fan commented:
I think we should have a config for it.

SparkQA commented:
**[Test build #108463 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108463/testReport)** for PR 25306 at commit [`220a9fb`](https://github.com/apache/spark/commit/220a9fb0cc854bf489e820e727af1b0de82958b8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108463/
Test PASSed.

advancedxy commented:
> I think we should have a config for it.

I could add a config for this. But is that really necessary?  Giving users the option to fallback to the old behaviour? 
We don't have a config for the read path, the write path should be the same eventually.

xuanyuanking commented:
The config is not only the option for users, but also a bandage when we found some cases missed in product env.

advancedxy committed:
Add convert inserting partitioned table config and it's enabled by default.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13725/
Test PASSed.

SparkQA commented:
**[Test build #108638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108638/testReport)** for PR 25306 at commit [`193a926`](https://github.com/apache/spark/commit/193a9261d66b3e89714bf2dadb67a0b5d77dda24).

SparkQA commented:
**[Test build #108638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108638/testReport)** for PR 25306 at commit [`193a926`](https://github.com/apache/spark/commit/193a9261d66b3e89714bf2dadb67a0b5d77dda24).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108638/
Test PASSed.

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala:
```diff
@@ -58,7 +58,11 @@ class HiveCommandSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
+    sql(
+      """
+        |CREATE TABLE parquet_tab4 (price int, qty int) partitioned by (year int, month int)
+        |STORED AS PARQUET
```
is it a necessary change?

advancedxy replied:
It's modified because I randomly chose it to make sure the insert into partitioned table can be safely converted.
However it should already covered in other test cases.

So it's neutral change. I can revert it if you think it's unnecessary.

advancedxy replied:
reverted

viirya commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala:
```diff
@@ -112,6 +112,15 @@ private[spark] object HiveUtils extends Logging {
+  val CONVERT_INSERTING_PARTITIONED_TABLE =
+    buildConf("spark.sql.hive.convertInsertingPartitionedTable")
+      .doc("When set to true, and \"spark.sql.hive.convertMetastoreParquet\" or " +
+        "\"spark.sql.hive.convertMetastoreOrc\" is true, the built-in ORC/Parquet writer is used" +
```
Usually we write like \`spark.sql.hive.convertMetastoreOrc\`?

advancedxy replied:
I referred `CONVERT_METASTORE_PARQUET_WITH_SCHEMA_MERGING`, will fix it.

viirya commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetMetastoreSuite.scala:
```diff
@@ -512,16 +512,14 @@ class HiveParquetMetastoreSuite extends ParquetPartitioningTest {
-    // Right now, insert into a partitioned Parquet is not supported in data source Parquet.
-    // So, we expect it is not cached.
-    assert(getCachedDataSourceTable(tableIdentifier) === null)
+    checkCached(tableIdentifier)
```
This change also affects ORC. Could you add test for Orc?

advancedxy replied:
I will check existing tests. ORC cases are not that much probably, I will add similar cases for ORC then.

viirya commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -209,13 +211,16 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
 
+          // Spark SQL's data source table now support static and dynamic partition insert. Source
+          // table converted from Hive table should always use dynamic.
+          val enableDynamicPartition = options.updated("partitionOverwriteMode", "dynamic")
```
I'm not sure about this part. Why it should always use dynamic?

advancedxy replied:
Please refer https://issues.apache.org/jira/browse/SPARK-20236.
TL;DR version: dynamic follows hive's style, static is the spark style which is different with hive. When converting InsertIntoHiveTable, we definitely want the hive behaviour. So it takes preference over `spark.sql.sources.partitionOverwriteMode`

advancedxy committed:
Address reviewers' comments.

SparkQA commented:
**[Test build #109107 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109107/testReport)** for PR 25306 at commit [`641f144`](https://github.com/apache/spark/commit/641f144579117d78ef2b5462a9c3bede86085af2).

advancedxy commented:
ping @cloud-fan and @viirya.
I pushed a new commit since no more comments are made.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/14172/
Test PASSed.

SparkQA commented:
**[Test build #109107 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109107/testReport)** for PR 25306 at commit [`641f144`](https://github.com/apache/spark/commit/641f144579117d78ef2b5462a9c3bede86085af2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/109107/
Test PASSed.

cloud-fan commented:
Can we update `DetermineTableStats` to catch `InsertIntoTable(HiveTableRelation)` as well? Similar to `ResolveRelations`.

advancedxy commented:
> Can we update `DetermineTableStats` to catch `InsertIntoTable(HiveTableRelation)` as well? 

will push a new commit later.

advancedxy committed:
DetermineTableStats now handles InsertIntoTable specially.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/14250/
Test PASSed.

SparkQA commented:
**[Test build #109179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109179/testReport)** for PR 25306 at commit [`c86a27b`](https://github.com/apache/spark/commit/c86a27b2a5e286733ad305de1d7e42d1373b3a3b).

SparkQA commented:
**[Test build #109179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109179/testReport)** for PR 25306 at commit [`c86a27b`](https://github.com/apache/spark/commit/c86a27b2a5e286733ad305de1d7e42d1373b3a3b).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/109179/
Test FAILed.

advancedxy commented:
@cloud-fan do you have any other suggestions?

The previous failure looks unrelated.

advancedxy commented:
Rebased with current master. ping @cloud-fan

advancedxy commented:
> Test build #109863 has finished for PR 25306 at commit 9899ade.
> 
> This patch fails Spark unit tests.
This patch merges cleanly.
This patch adds no public classes.

These test failures are introduced by this change. https://github.com/apache/spark/blob/9899adea2f125c350792d1d7b7ded3a629970766/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala#L146-L151

Once we enables the `InsertIntoTable` with stats,  the command executed in `InsertIntoHiveTable` as ` CommandUtils.updateTableStats(sparkSession, table)` will actually update stats, and invalid
a lot of assumptions.

Do we really want to calculate stats when inserting into tables?  @cloud-fan

cloud-fan commented:
@advancedxy this a good point. It looks to me that we mess up 2 things:
1. the table statistics from metastore
2. the data size statistics from HDFS

For table stats, we have special logic for it in places like `CommandUtils.updateTableStats`. However, in `DetermineTableStats`, we get size stats from HDFS and set it as table stats, which is incorrect.

My proposal is
```
class HiveTableRelation(..., statsOpt: Option[Statistics] = None) {
  ...
  override def computeStats(): Statistics = {
    tableMeta.stats.map(_.toPlanStats(output, conf.cboEnabled))
      .orElse(statsOpt)
      .getOrElse {
        throw new IllegalStateException("table stats must be specified.")
      }
  }
}
```

And `DetermineTableStats` should set `HiveTableRelation.statsOpt`

advancedxy commented:
@cloud-fan your proposal sounds good enough to me.

how would you like that to be addressed? 
1. Addressed in this pr.
2. Addressed in another Jira/pr first, and gathers others' opinion if any.

cloud-fan commented:
I think we can do it in this PR. Without this pr, the relation stats is useless if the relation is inside `InsertTable`

advancedxy committed:
Address comments.

SparkQA commented:
**[Test build #109954 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109954/testReport)** for PR 25306 at commit [`cd0d00b`](https://github.com/apache/spark/commit/cd0d00bb39e7ebbc8799f39cfba85275148927d2).

advancedxy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -136,10 +136,9 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+        // After SPARK-28573, sizeInBytes should be equal to dataSize.
         val relation1 = spark.table("t1").queryExecution.analyzed.children.head
-        assert(relation1.stats.sizeInBytes === spark.sessionState.conf.defaultSizeInBytes)
+        assert(relation1.stats.sizeInBytes === dataSize)
```
cc @wangyum, after changes applied to HiveTableRelation, we don't modify `tableMeta` any more. 
The converted `LogicalRelation(HadoopFsRelation, xx)` from `HiveTableRelation` will actual get size by `HadoopFsRelation.sizeInBytes`. 

Do you think this is an acceptable change or we should deal with stats differently in `DetermineTableStats`

cloud-fan replied:
looks better to me.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/14977/
Test PASSed.

SparkQA commented:
**[Test build #109954 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109954/testReport)** for PR 25306 at commit [`cd0d00b`](https://github.com/apache/spark/commit/cd0d00bb39e7ebbc8799f39cfba85275148927d2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/109954/
Test PASSed.

advancedxy commented:
> cc @wangyum, after changes applied to HiveTableRelation, we don't modify tableMeta any more.

Gently ping @wangyum and cc @cloud-fan

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala:
```diff
@@ -228,4 +229,64 @@ class HiveOrcQuerySuite extends OrcQueryTest with TestHiveSingleton {
+                   |CREATE TABLE dummy_orc_partitioned(key INT, value STRING)
+                   |PARTITIONED by (`date` STRING)
+                   |STORED AS ORC
+                   |LOCATION '${dir.toURI}'
```
we don't need to specify the table location, and save the `withTempPath`.

advancedxy committed:
Address comments.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15029/
Test PASSed.

SparkQA commented:
**[Test build #110009 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110009/testReport)** for PR 25306 at commit [`b80166b`](https://github.com/apache/spark/commit/b80166badf5f63d0dffa7dd9eee57c0ff7b6c9f0).

SparkQA commented:
**[Test build #110009 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110009/testReport)** for PR 25306 at commit [`b80166b`](https://github.com/apache/spark/commit/b80166badf5f63d0dffa7dd9eee57c0ff7b6c9f0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110009/
Test PASSed.

cloud-fan commented:
thanks, merging to master!
